### DISCUSSION TITLE
A different spectrogram plugin

### DISCUSSION THREAD

**Discussion by danijel3** [katspaugh/wavesurfer.js#3916] [Ideas]
I've tried using the provided spectrogram plugin, but found it a bit lacking. Maybe I could spend more time fine-tuning it and fixing some bugs, but I found a different solution.

The EMU project has a beautiful spectrogram that is very similar to what "pros" use for doing phonetics in programs like [Praat](https://www.fon.hum.uva.nl/praat/) (the de-fact standard for research in phonetics). It also provides a slightly different philosophy to drawing the spectrogram. Instead of hogging the main thread for processing large files, it draws the spectrogram only for the visible chunk of audio and does so using a background worker instead of the main thread.

You can view it by visiting https://ips-lmu.github.io/EMU-webApp/ and choosing a demo from the button on the top of the interface. Scrolling through the audio is done using 'w','a','s','d' on the keyboard. You can find other neat keyboard and UI tricks in the documentation accessible by clicking the emu logo on the top left of the screen.

I really dig EMU and think it is (was) a great project. The only problem (why I can't use it myself) is it's an Angular app and not a simple component I can embed everywhere.

I managed to adapt their spectrogram worker to wavesurfer in this fork: https://github.com/danijel3/wavesurfer.js

I don't claim coming up with this method and it was fully adapted from the original project (https://github.com/IPS-LMU/EMU-webApp). The worker is a 1-to-1 copy from their repo. I just made the plugin that adapts the worker and documented as much as I could.

![Screenshot from 2024-10-19 14-57-00](https://github.com/user-attachments/assets/882250f8-a5f1-4b23-b768-2dc0622b7105)

What do you suggest I do with this? Can we add it to wavesurfer or is there something else we need to do?

**Comment by katspaugh** [comment]
Great idea! That repo seems to have a MIT license so I guess it’s fine to just copy that worker code and add a comment on top crediting them.

**Reply by danijel3** [reply]
So you think it's okay to make a PR, or are 2 spectrogram plugins too much? I understand the original plugin is meant to simulate Audacity and some people might be used to it (especially the color scheme). This one is more meant to emulate phonetic analysis tools. Audacity is more for music and sound production.

Alternatively, I can simply set up a repo like the one you did for multitrack editing and just maintain the code there...

**Reply by katspaugh** [reply]
Hey @danijel3, sorry I missed your message!

I think setting up your own repo like we did for multitrack would be a great option. We could link it in the plugin list here as an alternative option.

We could eventually phase out the existing spectrogram plugin once yours is confirmed to have feature parity with it.

**Reply by danijel3** [reply]
Hey, I made the separate repo: https://github.com/danijel3/wavesurfer-spectrogram-emu

**Comment by melijimenez11** [comment]
@katspaugh @danijel3 

I'm interested in this, as I'm using wavesurfer to analyze bioacoustic sound data. the EMU version seems more appropriate. I'm a beginner sort of. Can you explain how I would go about incorporating this into my project? Or can it be added as an option to wavesurfer?

**Reply by danijel3** [reply]
Sorry it took so long. First holidays, then vacation...

You'll find everything here: https://github.com/danijel3/wavesurfer-spectrogram-emu
